raft topology: implement upgrade and recovery procedure

piodul opened the pull request:
This PR implements a procedure that upgrades existing clusters to use raft-based topology operations. The procedure does not start automatically, it must be triggered manually by the administrator after making sure that no topology operations are currently running.

Upgrade is triggered by sending `POST /storage_service/raft_topology/upgrade` request. This causes the topology coordinator to start who drives the rest of the process: it builds the `system.topology` state based on information observed in gossip and tells all nodes to switch to raft mode. Then, topology coordinator runs normally.

Upgrade progress is tracked in a new static column `upgrade_state` in `system.topology`.

The procedure also serves as an extension to the current recovery procedure on raft. The current recovery procedure requires restarting nodes in a special mode which disables raft, perform `nodetool removenode` on the dead nodes, clean up some state on the nodes and restart them so that they automatically rebuild the group 0. Raft topology fits into existing procedure by falling back to legacy topology operations after disabling raft. After rebuilding the group 0, upgrade needs to be triggered again.

Because upgrade is manual and it might not be convenient for administrators to run it right after upgrading the cluster, we allow the cluster to operate in legacy topology operations mode until upgrade, which includes allowing new nodes to join. In order to allow it, nodes now ask the cluster about the mode they should use to join before proceeding by using a new `JOIN_NODE_QUERY` RPC.

The procedure is explained in more detail in `topology-over-raft.md`.

Fixes: https://github.com/scylladb/scylladb/issues/15008

piodul commented:
TODOs that I'd like to implement in the near future, but are perhaps not critical and should not block review:

- Use get_endpoint_states (shadow round verb) in the topology coordinator to collect nodes' gossiper information
- Forbid modify_group0_config verb when upgrade starts
- Perform more validity checks on the gossiper state when topology coordinator builds the initial state
- Add a test which simulates the upgrade procedure getting stuck and checks that it is possible to get out of it by performing recovery
- Consider using gossip shadow round instead of `JOIN_NODE_QUERY` to determine the way a node should attempt joining the cluster

piodul commented on service/storage_service.hh:
```diff
@@ -750,8 +750,15 @@ private:
         // The node uses raft-based topology operations
         raft
     };
+    // The _topology_change_kind_enabled variable is first initialized in `join_cluster`.
+    // After the node successfully joins, the control over the variable is yielded
+    // to `topology_state_load`, so that it can control it during the upgrade from gossiper
+    // based topology to raft-based topology.
+    bool _manage_topology_change_kind_from_group0 = false;
```
This is a hack to work around the following issue: when a node creates a new cluster, it starts group 0 with an empty state. If `topology_state_load` were to always update this field then it would temporarily revert `_topology_change_kind_enabled` to `legacy`.

I'll try to fix this by initializing the newly created group 0 during `setup_group0` (effectively move `raft_initialize_discovery_leader` to happen during it through the handshaker).

I'll also try to reorganize the beginning of `join_token_ring` so that the mode we use is implied by the control flow and not `_topology_change_kind_enabled`, at least not until gossiping starts (because it reads the flag).

piodul replied:
Gleb's arguments in the other thread convinced me that it's not worth implementing the blocking semantics for `upgrade` (needs an additional step), so I'm resolving.

piodul replied:
I added a FIXME comment to the code.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6238/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6238/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7537/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6238/)
#### Failed Tests (6/24198):
- [test_topology_upgrade_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6238/testReport/junit/%28root%29/test_topology_upgrade/Tests___Unit_Tests___test_topology_upgrade_basic) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_basic)
- [test_topology_upgrade_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6238/testReport/junit/%28root%29/test_topology_upgrade/Tests___Unit_Tests___test_topology_upgrade_basic_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_basic)
- [test_topology_upgrade_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6238/testReport/junit/%28root%29/test_topology_upgrade/Tests___Unit_Tests___test_topology_upgrade_basic_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_basic)
- [topology_custom.test_topology_upgrade.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6238/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_topology_upgrade_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_topology_upgrade.release.1)
- [topology_custom.test_topology_upgrade.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6238/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_topology_upgrade_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_topology_upgrade.release.2)
- [topology_custom.test_topology_upgrade.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6238/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_topology_upgrade_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_topology_upgrade.release.3)
#### Build Details:
- Duration: 2 hr 2 min
- Builder: spider1.cloudius-systems.com

piodul commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6238/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6238/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7537/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6238/)
> #### Failed Tests (6/24198):
> 
>     * [test_topology_upgrade_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6238/testReport/junit/%28root%29/test_topology_upgrade/Tests___Unit_Tests___test_topology_upgrade_basic) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_basic)
> 
>     * [test_topology_upgrade_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6238/testReport/junit/%28root%29/test_topology_upgrade/Tests___Unit_Tests___test_topology_upgrade_basic_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_basic)
> 
>     * [test_topology_upgrade_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6238/testReport/junit/%28root%29/test_topology_upgrade/Tests___Unit_Tests___test_topology_upgrade_basic_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_basic)
> 
>     * [topology_custom.test_topology_upgrade.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6238/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_topology_upgrade_release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_topology_upgrade.release.1)
> 
>     * [topology_custom.test_topology_upgrade.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6238/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_topology_upgrade_release_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_topology_upgrade.release.2)
> 
>     * [topology_custom.test_topology_upgrade.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6238/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_topology_upgrade_release_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_topology_upgrade.release.3)
> 
> 
> #### Build Details:
> 
>     * Duration: 2 hr 2 min
> 
>     * Builder: spider1.cloudius-systems.com

The test that failed uses an error injection and is not yet skipped in `release` mode. Will fix.

The test was run 8 times but it only failed 3 times. The test uses an error injection to start the cluster in legacy mode and then checks that it indeed didn't upgrade to raft topology. Given the circumstances, the test should have failed all 8 times - this implies that new clusters can sometimes be accidentally started in legacy mode. I'll investigate this bug.

piodul commented:
> The test that failed uses an error injection and is not yet skipped in `release` mode. Will fix.
> 
> The test was run 8 times but it only failed 3 times. The test uses an error injection to start the cluster in legacy mode and then checks that it indeed didn't upgrade to raft topology. Given the circumstances, the test should have failed all 8 times - this implies that new clusters can sometimes be accidentally started in legacy mode. I'll investigate this bug.

False alarm. The test was actually run 9 times, 3 times in each mode (debug, dev, release). All runs in release mode failed, looks like there is no bug, the only conclusion is that I can neither count nor read properly.

kbr-scylla commented on test/topology_custom/test_topology_recovery_basic.py:
```diff
@@ -0,0 +1,80 @@
+    hosts = await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
+
+    logging.info(f"Restarting hosts {hosts} in recovery mode")
+    await asyncio.gather(*(enter_recovery_state(cql, h) for h in hosts))
```
Now we could do this part using maintenance mode -- no need to write to the scylla_local table to go into recovery, but start with maintenance mode config flag (which automatically triggers raft recovery too). I think that would be a nice practical test for maintenance mode.

(We can switch in a follow-up, and the existing recovery tests too, if it's a hassle)

piodul replied:
Let's do it in a followup

kbr-scylla commented on test/topology_custom/test_topology_recovery_basic.py:
```diff
@@ -0,0 +1,80 @@
+        assert status == "not_upgraded"
+
+    logging.info("Waiting until all nodes see others as alive")
+    await asyncio.gather(*(manager.server_sees_others(srv.server_id, len(servers) - 1, time.time() + 60) for srv in servers))
```
`servers_see_each_other`

kbr-scylla commented on test/topology_custom/test_topology_recovery_basic.py:
```diff
@@ -0,0 +1,80 @@
+    await asyncio.gather(*(wait_until_topology_upgrade_finishes(manager, h.address, time.time() + 60) for h in hosts))
+
+    logging.info("Checking consistency of data in system.topology and system.cdc_generations_v3")
+    await asyncio.gather(*(check_system_topology_and_cdc_generations_v3_consistency(manager, h) for h in hosts))
```
Perhaps we could verify that the state is the same across nodes? (Which should be true without having to do read barriers from the test, as there is the global barrier done by coordinator already) Just checking if fields are `not None` doesn't seem to provide much value.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -1681,13 +1682,16 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
             try {
                 co_await me.track_upgrade_progress_to_topology_coordinator(sys_dist_ks);
+                if (gossip_feature_enabler) {
+                    co_await me._gossiper.unregister_(gossip_feature_enabler);
```
Isn't there a race here? Looks like the running of feature enabled overlaps a bit with the topology coordinator. So it may call enable_features() once or twice before we disable it. Not sure if it's a problem. Perhaps a correct solution would have enable_features() itself check where in the upgrade we are

piodul replied:
I can change it, thoguh I don't think the current code will be a problem. Features that are initially put into `system.topology` are completely the same as the features that are currently gossiped. `enable_features` will overwrite some state that is replicated from raft, but it will overwrite it with the same value (enabled features in `system.scylla_local`).

kbr-scylla replied:
Imagine a day passes between these two lines. Is it still correct? What if, let's say, topology coordinator enables some new feature during that time (not sure if possible)?

piodul replied:
> not sure if possible

That would require doing a rolling upgrade to a new version with new features, and after restart this code path wouldn't be touched at all.

Anyway, I get the point, we don't want to try to look at this code every time and have convince ourselves that it works. I will change the feature enabler to check the flag itself.

kbr-scylla commented on service/raft/group0_state_machine.cc:
```diff
@@ -43,11 +43,19 @@
+group0_state_machine::group0_state_machine(raft_group0_client& client, migration_manager& mm, storage_proxy& sp, storage_service& ss, raft_address_map& address_map, gms::feature_service& feat, bool topology_change_enabled)
+        : _client(client), _mm(mm), _sp(sp), _ss(ss), _address_map(address_map), _topology_change_enabled(topology_change_enabled)
+        , _topology_on_raft_support_listener(feat.supports_consistent_topology_changes.when_enabled([this] () noexcept {
+            _topology_change_enabled = true;
```
But by the time the first pull may be triggered, already requiring topology to be included --- let's say that the topology coordinator has committed the first group 0 command as part of the upgrade procedure --- we might still not have noticed that the feature is enabled, no? In which case we'll skip pulling topology, and that's bad.

piodul replied:
Hmm, right. Rookie mistake from my part - assuming that the feature is enabled on other nodes if it is enabled on the local node. Before we start the upgrade, we need to somehow make sure that other nodes will, from that moment on, pull the topology snapshot...

piodul replied:
In the end, it will be the responsibility of the administrator to wait until the feature becomes enabled on all nodes. I mention this in the `topology-over-raft.md` doc and added a comment in the code about the improper use of the cluster features.

kbr-scylla replied:
Please also add a comment in the documentation issue:
https://github.com/scylladb/scylladb/issues/15934

piodul commented on service/topology_coordinator.cc:
```diff
@@ -2008,6 +2016,11 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
     // Returns true if the state machine was transitioned into tablet migration path.
     future<bool> maybe_start_tablet_migration(group0_guard);
 
+    // Returns true if it successfully moved the upgrade forward.
```
Good point. In some previous version of the code I used to start topology coordinator before upgrade could start, so in case of `not_started` I returned `false` - but now this should not happen and there is an `on_internal_error. I'll remove the bool from the return type.

piodul commented on service/topology_coordinator.cc:
```diff
@@ -2066,6 +2079,137 @@ future<bool> topology_coordinator::maybe_start_tablet_migration(group0_guard gua
+    for (const auto& host_id: tmptr->get_all_endpoints()) {
+        const auto ep = tmptr->get_endpoint_for_host_id_if_known(host_id);
+        if (!ep) {
+            throw std::runtime_error(format("IP of node with ID {} is not known", host_id));
```
Yes, in case of an otherwise unhandled exception the topology coordinator will retry (see `topology_coordinator::run`).

piodul commented on service/storage_service.cc:
```diff
@@ -5209,6 +5217,29 @@ future<join_node_request_result> storage_service::join_node_request_handler(join
     join_node_request_result result;
     rtlogger.info("received request to join from host_id: {}", params.host_id);
 
+    if (legacy_topology_change_enabled()) {
```
Yes, this should not happen. I'll highlight this in a comment/error message.

kbr-scylla replied:
If something indicates a bug, I recommend `on_internal_error`

piodul commented on service/storage_service.cc:
```diff
@@ -1110,6 +1110,59 @@ future<> storage_service::update_topology_with_local_metadata(raft::server& raft
+        if (_topology_state_machine._topology.ustate == topology::upgrade_state::done) {
+            throw std::runtime_error("The cluster is already upgraded to use topology on raft.");
+        } else if (_topology_state_machine._topology.ustate != topology::upgrade_state::not_upgraded) {
+            throw std::runtime_error("The upgrade has already started and is in progress.");
```
Definitely makes sense

piodul commented on service/storage_service.cc:
```diff
@@ -398,6 +398,7 @@ future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tm
         // Also ip -> id mapping is needed for address map recreation on reboot
         if (is_me(host_id)) {
             co_await _sys_ks.local().update_tokens(rs.ring.value().tokens);
+            co_await set_gossip_tokens(_gossiper, rs.ring.value().tokens, _topology_state_machine._topology.current_cdc_generation_id);
```
Truncating `system.cdc_generations_v3` is not done during recovery so sounds like it could work, I'll try it.

piodul commented on service/topology_coordinator.cc:
```diff
@@ -2066,6 +2079,137 @@ future<bool> topology_coordinator::maybe_start_tablet_migration(group0_guard gua
+    guard = std::move(guard_);
+
+    const bool add_ts_delay = true; // This is not the first generation, so add delay
+    auto cdc_gen_ts = cdc::new_generation_timestamp(add_ts_delay, _ring_delay);
```
Between `new_generation_timestamp` and `update_topology_state` that sets the `current_cdc_generation` field there is no preemption, so the delay should be microscopic. I can move this invocation later to make it more obvious, though.

piodul commented on service/storage_service.cc:
```diff
@@ -2449,6 +2446,50 @@ future<> storage_service::join_cluster(sharded<db::system_distributed_keyspace>&
+
+        if (_group0->load_my_id() == g0_info.id) {
+            // We're creating the group 0.
+            // If we are already bootstrapped, then it means that we are restarting after recovery and we should continue in legacy mode.
```
Yeah, I tend be lazy and shorten "raft topology mode" to "raft mode", but I agree it's ambiguous - I'll go over the places where I use the short form and will clarify.

piodul commented on docs/dev/topology-over-raft.md:
```diff
@@ -362,3 +368,75 @@ In case of failures like timeouts, connection issues, etc., the topology
+  and an appropriate message about the cluster being ready for upgrade is printed
+- Trigger the upgrade via `POST /storage_service/raft_topology/upgrade` HTTP route
+- Monitor progress of the upgrade via `GET /storage_service/raft_topology/upgrade`
+  or via observing the logs
```
> Maybe the POST should block until upgrade is finished?

Perhaps. Should it wait until the current node reaches `done`, or all nodes reach `done`? As currently described in `topology-on-raft.md`, it is necessary to check that all nodes reached `done` state, otherwise if you only check one then some nodes might still be in `final_global_barrier`.

> Although I'm guessing this also waits for the group0/schema part of the upgrade procedure (if applicable) (I recall we briefly discussed this at some point)

Yes, it does. This is the first thing that `build_coordinator_state` does.

piodul replied:
Gleb's arguments in the other thread convinced me that it's not worth implementing the blocking semantics for `upgrade` (needs an additional step), so I'm resolving.

gleb-cloudius commented on service/raft/join_node.hh:
```diff
@@ -14,6 +14,14 @@
+struct join_node_query_result {
+    // If true, the node should use raft topology.
+    // Otherwise, it should fall back to legacy node operations.
+    bool use_raft_topology;
```
May be enum? Who knows if we will need to introduce some other mode.

gleb-cloudius commented on service/topology_coordinator.cc:
```diff
@@ -2066,6 +2079,137 @@ future<bool> topology_coordinator::maybe_start_tablet_migration(group0_guard gua
+        co_await build_coordinator_state(std::move(guard));
+        co_return true;
+
+    case topology::upgrade_state::final_global_barrier: {
```
Why do we need this final global barrier? What is it waits for?

kbr-scylla replied:
> given that the administrator has to perform a "global barrier" themselves by checking the upgrade state on all node.

The coordinator may automatically start some work after this barrier, e.g. suppose we add further "upgrade" steps for auth, service levels, tablets...

But maybe those steps will start with a barrier

Although this barrier definitely doesn't hurt. It's also needed if we want the POST to block

gleb-cloudius replied:
> > given that the administrator has to perform a "global barrier" themselves by checking the upgrade state on all node.
> 
> The coordinator may automatically start some work after this barrier, e.g. suppose we add further "upgrade" steps for auth, service levels, tablets...
> 
> But maybe those steps will start with a barrier

Everything that is driven by the coordinator goes through raft cmd rpc which runs barrier to see the latest state.
Explicit one is not needed usually.

> 
> Although this barrier definitely doesn't hurt. It's also needed if we want the POST to block

We will need another barrier after we move to "done" state then :) Lets not do barriers we do not need (yes they are harmless, but just complicate code unnecessary).

gleb-cloudius commented on service/topology_coordinator.cc:
```diff
@@ -2066,6 +2079,137 @@ future<bool> topology_coordinator::maybe_start_tablet_migration(group0_guard gua
+    co_await _group0.wait_for_all_nodes_to_finish_upgrade(_as);
+
+    // Execute a global barrier to make sure that all nodes see
+    // the build_coordinator_state upgrade_state
```
Why all nodes have to see the state before we proceed with the upgrade?

piodul replied:
After the barrier, the whole cluster will be put into a state where no legacy topology operations are allowed anymore. This prevents a situation where some nodes allow initiating legacy topology operations and some others already operate in raft mode.

To be fair... Given that we already require administrators to abstain from issuing topology requests during the procedure - and they need to check that all nodes ultimately see the "done" state - this doesn't seem to bring much value, so maybe this precaution can be dropped. On the other hand, it doesn't cost us much to do it (no additional states, RPC verbs etc.). @kbr-scylla WDYT?

gleb-cloudius replied:
> To be fair... Given that we already require administrators to abstain from issuing topology requests during the procedure - and they need to check that all nodes ultimately see the "done" state - this doesn't seem to bring much value

Exactly.

kbr-scylla replied:
Such precautions don't hurt, and they will be needed once we make the procedure fully automatic and safe from administrator errors (that is the plan, remember?) so why not add it now

gleb-cloudius replied:
Because we cannot yet know where they are needed and for what. It may be not enough or not needed here.

kbr-scylla replied:
We do know because we started by designing the full procedure

@piodul isn't this one obvious? We know we'll need it. We want all nodes to learn that we're in building phase to disable topology/schema changes

piodul replied:
We don't need it right now, and adding it later shouldn't be a problem.

Right now it makes people trying to understand the current procedure, but who are not familiar with the full procedure, scratch their head.

gleb-cloudius replied:
Lets add things when they are needed. This way we will know why  they are needed.

piodul commented on service/topology_coordinator.cc:
```diff
@@ -2187,7 +2331,7 @@ future<> topology_coordinator::run() {
             }
 
-            bool had_work = co_await handle_topology_transition(std::move(guard));
+            bool had_work = co_await do_work(std::move(guard));
```
Sure, I can inline if it makes it clearer.

piodul commented on service/topology_coordinator.cc:
```diff
@@ -440,15 +441,13 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                } catch (...) {
+                    std::throw_with_nested(std::runtime_error(::format(
+                        "raft topology: make_new_cdc_generation_data: get_sharding_info:"
+                        " can't get sharding info for node {}, owner of token {}", *ep, end)));
```
Hm, right, that was not intentional for the existing cases (i.e. when the function is called under other circumstances than upgrade) as it uses info from topology which should be consistent.

I'd rather not trigger an internal error during upgrade - if something is wrong and the gossip state is broken it's better to throw an exception and retry rather than: 1) crash the current node 2) elect a new leader and start the topology coordinator on it 3) goto 1.

I'll try to reorganize so that it throws in the right circumstances and causes an internal error in the remaining ones.

gleb-cloudius replied:
Lets ask @kbr-scylla if it have to be internal_error in non upgrade case and what will happen if we throw instead.

kbr-scylla replied:
If the information is missing in system.topology (non upgrade case), it's a bug, and bugs are marked with on_internal_error.

Reminder: on_internal_error doesn't crash in production, unless someone sets `--abort-on-internal-error` (not recommended), but it will crash in tests, which is good to catch issues early especially in test.py which doesn't detect ERRORs in logs.

piodul replied:
> Reminder: on_internal_error doesn't crash in production

Right, I forgot about that. In that case it makes sense to go with it in both cases.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1608,6 +1608,50 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
     assert(_group0);
     co_await _group0->finish_setup_after_join(*this, _qp, _migration_manager.local(), _raft_experimental_topology);
     co_await _cdc_gens.local().after_join(std::move(cdc_gen_id));
+
```
> Add a fiber which is spawned only in legacy mode that will wait until:
>
> - The schema-on-raft upgrade finishes,
> - The SUPPORTS_CONSISTENT_CLUSTER_MANAGEMENT feature is enabled,
> - The upgrade is triggered by the user.

Do we need a fiber to track all of this or each one has some kind of event that can be used to trigger the raf monitor fiber creation?

piodul replied:
Of those three, I think that only `supports_consistent_topology_changes` allows that through `gms::feature::when_enabled`. Others (`wait_until_group0_upgraded` and waiting for topology state machine to enter the right state) requires a fiber that waits for the right events.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1608,6 +1608,50 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
+
+    // Start the topology coordinator monitor fiber. If we are the leader, this will start
+    // the topology coordinator which is responsible for driving the upgrade process.
+    _raft_state_monitor = raft_state_monitor_fiber(_group0->group0_server(), sys_dist_ks);
```
What happens if the fiber throws before it gets here?

piodul replied:
The only exception that can be thrown is `abort_requested_exception`  when waiting on any of the conditions is aborted due to the node shutting down - this was intentional. Any other exception is unexpected.

I can add a `try..catch` and log a message if an unexpected exception occurs.

gleb-cloudius replied:
> The only exception that can be thrown is abort_requested_exception

Are you sure? It is a lot of code there.

piodul replied:
OK, I am not 100% sure. However, the code in `track_upgrade_progress_to_topology_coordinator` before this line just waits for some conditions to happen; the waits are abortable and can throw `abort_requested_exception`. It's not a part of their interface to throw other exceptions.

Of course, there is a risk that some of them may throw `std::bad_alloc` or something similarly unexpected (`when_enabled()` uses `make_lw_shared` so it technically could). Do you suggest to handle them in some other way than printing the error?

gleb-cloudius replied:
It is better to either restart the fiber or kill the Scylla since without the fiber it will not work properly.

kbr-scylla replied:
>  It is better to either restart the fiber or kill the Scylla since without the fiber it will not work properly.


Kill Scylla? We decided not to kill Scylla when group 0 server dies and for a good reason -- the data path may still work and killing Scylla would only hurt client applications.

In the same spirit we should not kill Scylla if topology coordinator fiber dies. It will only block cluster metadata operations but the data path may be functioning.

But I'm not sure if there is any reason (except shutdown and leader stepdown) for the topology coordinator fiber not to run forever

gleb-cloudius replied:
If you remember there we complains already about our decision to not kill Scylla if group0 dies. When only schema changes worked through it it made some sense, but now it just makes a lot of thing non functional without easy way to understand why.

kbr-scylla replied:
> If you remember there we complains already about our decision to not kill Scylla if group0 dies.

I don't remember complaints, I remember questions. If there was a complaint then it came from a person who didn't understand why such decision was made, and the complaint stopped once we explained the reason.

> When only schema changes worked through it it made some sense, but now it just makes a lot of thing non functional without easy way to understand why.

If it's not easy to understand it's a documentation issue.

We should not needlessly disrupt client applications.
The choice is between:
- a bug kills the entire Scylla cluster, all apps stop working, millions of $$ loss, customer is furious, you get paged at night
- a bug kills topology coordinator / group 0 server on all nodes, but applications are still working. You sleep through the night and investigate the issue in the morning. No $$ loss

a simple `bad_alloc` is all it takes

The choice is obvious

kbr-scylla replied:
> a simple bad_alloc is all it takes

And I'm mentioning `bad_alloc` because we already saw them killing group 0 server. So it doesn't even need to be a "bug" per se

kbr-scylla replied:
What we can do to improve our testing is to call `on_internal_error` if group 0 server dies, for example. It won't cause production clusters to crash (unless client likes to play Russian roulette and uses `--abort-on-internal-error` which is not recommended, they asked for it) but it might make tests detect issues faster.

gleb-cloudius replied:
> > If you remember there we complains already about our decision to not kill Scylla if group0 dies.
> 
> I don't remember complaints, I remember questions. If there was a complaint then it came from a person who didn't understand why such decision was made, and the complaint stopped once we explained the reason.
> 

Was from @tgrabiec iirc.

> > When only schema changes worked through it it made some sense, but now it just makes a lot of thing non functional without easy way to understand why.
> 
> If it's not easy to understand it's a documentation issue.

Documentation solves nothing here. Nobody is going to read a documentation at three o'clock to understand why a node cannot be replaced. And fixing a failed group0 meas you need to restart a node and I am sure admins will be happy to do it when they already missing one node.

> 
> We should not needlessly disrupt client applications. The choice is between:
> 
>     * a bug kills the entire Scylla cluster, all apps stop working, millions of $$ loss, customer is furious, you get paged at night
> 
>     * a bug kills topology coordinator / group 0 server on all nodes, but applications are still working. You sleep through the night and investigate the issue in the morning. No $$ loss
> 
> 
> a simple `bad_alloc` is all it takes

There are bad_alloc already that will kill a node. And bad_alloc usually do not kill entire cluster.

> 
> The choice is obvious

Not that obvious as you want it to be. Some problems better to fix as fast as possible and not wait for situation to aggravate. 

But I proposed to alternative. Another was to restart the fiber. Printing an error in the logs and continue as nothing happens is the worst of all the options.

kbr-scylla replied:
> Documentation solves nothing here. Nobody is going to read a documentation at three o'clock to understand why a node cannot be replaced.

Documentation consists not only of internet pages, but also on good log messages, good errors returned from APIs, good metrics and so on.

The choice is between:
- reading documentation at 3 o'clock while you can't replace a node 
- reading documentation at 3 o'clock while none of your applications are working at all because your whole DB crashed and you're losing $$ every second

> And fixing a failed group0 meas you need to restart a node and I am sure admins will be happy to do it when they already missing one node.

I'm sure admins will be even more happy if we do the shutdown part of that restart automatically for them, which is basically what you propose. And chance is on all nodes at the same time! (While they can do a rolling restart)

> And bad_alloc usually do not kill entire cluster.

Unless it's caused by applying some logic that is coincidentally being applied on every node due to a state machine replication mechanism

> Not that obvious as you want it to be. Some problems better to fix as fast as possible and not wait for situation to aggravate.

Group 0 server dying is a serious issue I agree, but that's why we have monitoring and **alerts**, and serious alerts from customers are already often a reason for Support getting called. In general if there's some serious trouble with node we already have a way to make everyone aware who needs to be, you don't need to make the situation even more serious than it is.

And we don't only print error in the logs, we also have the metric with group 0 status which should be used for triggering alerts (and if we don't do it, we should, @kostja was this ever communicated outside core team?)

Restarting the fiber is also an option, we can reach consensus there.

tgrabiec replied:
> > > If you remember there we complains already about our decision to not kill Scylla if group0 dies.
> > 
> > 
> > I don't remember complaints, I remember questions. If there was a complaint then it came from a person who didn't understand why such decision was made, and the complaint stopped once we explained the reason.
> 
> Was from @tgrabiec iirc.
> 

Not sure which discussion you mean, maybe this one: https://github.com/scylladb/scylladb/issues/16493#issuecomment-1878762701
I was against aborting the node on failure to apply a command.

kbr-scylla replied:
> Restarting the fiber is also an option, we can reach consensus there.

Although we should probably still trigger an alert if the situation is not resolved (e.g. we see that we keep restarting the fiber multiple times within a small time period)

gleb-cloudius replied:
> > > > If you remember there we complains already about our decision to not kill Scylla if group0 dies.
> > > 
> > > 
> > > I don't remember complaints, I remember questions. If there was a complaint then it came from a person who didn't understand why such decision was made, and the complaint stopped once we explained the reason.
> > 
> > 
> > Was from @tgrabiec iirc.
> 
> Not sure which discussion you mean, maybe this one: [#16493 (comment)](https://github.com/scylladb/scylladb/issues/16493#issuecomment-1878762701) I was against aborting the node on failure to apply a command.

It was not about group0 IIRC but topology coordinator fiber. In the end we made it noexcept.

kbr-scylla replied:
> We do not discuss group0 zero here. No need to attack a straw man you made here for yourself.

I'm not arguing with a strawman but with what you said:

> If you remember there we complains already about our decision to not kill Scylla if group0 dies. When only schema changes worked through it it made some sense, but now it just makes a lot of thing non functional without easy way to understand why.

That was about group 0 server, no?

gleb-cloudius replied:
> > If you remember there we complains already about our decision to not kill Scylla if group0 dies. When only schema changes worked through it it made some sense, but now it just makes a lot of thing non functional without easy way to understand why.
> 
> That was about group 0 server, no?

Yes, but here we discuss a fiber that does not have any global effect, so why go on the tangent. 

With group0 there is a potential to have a bug that may kill all nodes, but even there there are cases where you can tell that it is just a node local out of memory. And nobody wants to micromanage cluster nodes. If a restart may improve the situation 10 out of 10 admins will chose restart instead of paging. Of course the best is to have an integration with a cluster management so that a node may ask the manager to be restarted and then if a manager gets to much of those requests it will do it one by one, but this is not something we have now.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -1608,6 +1608,50 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
+    // the topology coordinator which is responsible for driving the upgrade process.
+    _raft_state_monitor = raft_state_monitor_fiber(_group0->group0_server(), sys_dist_ks);
+
+    // Wait until upgrade is finished
```
What for? raft monitor is already started and this fiber has nothing else to do after that, no?

piodul replied:
The fiber waits for upgrade to finish in order to print the message about it. If not for that, the fiber could exit here, yes.

piodul commented:
v2.1: rebased

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6485/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6485/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6485/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7724/)
#### Failed Tests (6/24480):
- [test_topology_upgrade_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6485/testReport/junit/%28root%29/test_topology_upgrade/Tests___Unit_Tests___test_topology_upgrade_basic) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_basic)
- [test_topology_upgrade_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6485/testReport/junit/%28root%29/test_topology_upgrade/Tests___Unit_Tests___test_topology_upgrade_basic_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_basic)
- [test_topology_upgrade_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6485/testReport/junit/%28root%29/test_topology_upgrade/Tests___Unit_Tests___test_topology_upgrade_basic_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_basic)
- [topology_experimental_raft.test_topology_upgrade.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6485/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_upgrade_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_upgrade.release.1)
- [topology_experimental_raft.test_topology_upgrade.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6485/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_upgrade_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_upgrade.release.2)
- [topology_experimental_raft.test_topology_upgrade.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6485/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_upgrade_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_upgrade.release.3)
#### Build Details:
- Duration: 1 hr 57 min
- Builder: spider6.cloudius-systems.com

piodul commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6485/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6485/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6485/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7724/)
> 
> #### Failed Tests (6/24480):
> * [test_topology_upgrade_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6485/testReport/junit/%28root%29/test_topology_upgrade/Tests___Unit_Tests___test_topology_upgrade_basic) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_basic)
> * [test_topology_upgrade_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6485/testReport/junit/%28root%29/test_topology_upgrade/Tests___Unit_Tests___test_topology_upgrade_basic_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_basic)
> * [test_topology_upgrade_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6485/testReport/junit/%28root%29/test_topology_upgrade/Tests___Unit_Tests___test_topology_upgrade_basic_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_upgrade_basic)
> * [topology_experimental_raft.test_topology_upgrade.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6485/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_upgrade_release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_upgrade.release.1)
> * [topology_experimental_raft.test_topology_upgrade.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6485/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_upgrade_release_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_upgrade.release.2)
> * [topology_experimental_raft.test_topology_upgrade.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6485/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_upgrade_release_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_upgrade.release.3)
> 
> #### Build Details:
> * Duration: 1 hr 57 min
> * Builder: spider6.cloudius-systems.com

Same issue as the previous failure in https://github.com/scylladb/scylladb/pull/17077#issuecomment-1918885737 ... I first fixed it by skipping the test in `topology_custom/suite.yaml`, but then I moved the tests to `topology_experimental_raft` and forgot to mark it as ignored there. Will fix soon

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6492/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6492/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7732/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6492/)
#### Failed Tests (1/24474):
- [test_replay_after_schema_change](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6492/testReport/junit/batch_test/TestBatch/Tests___dtest___test_replay_after_schema_change) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replay_after_schema_change)
#### Build Details:
- Duration: 2 hr 3 min
- Builder: spider2.cloudius-systems.com

piodul commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6492/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6492/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7732/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6492/)
> 
> #### Failed Tests (1/24474):
> * [test_replay_after_schema_change](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6492/testReport/junit/batch_test/TestBatch/Tests___dtest___test_replay_after_schema_change) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replay_after_schema_change)

https://github.com/scylladb/scylladb/issues/17192

> 
> #### Build Details:
> * Duration: 2 hr 3 min
> * Builder: spider2.cloudius-systems.com

piodul commented:
@gleb-cloudius @kbr-scylla gentle review ping

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -2457,6 +2454,50 @@ future<> storage_service::join_cluster(sharded<db::system_distributed_keyspace>&
+        _topology_change_kind_enabled = topology_change_kind::legacy;
+    } else if (_group0->joined_group0()) {
+        slogger.info("Restarting in raft mode");
+        _topology_change_kind_enabled = topology_change_kind::raft;
```
But can't it be "schema over raft" only mode?

piodul replied:
Hm, right. At this point of this PR (i.e. right after this commit) it is true that this piece of code is incorrect. It is later fixed by `storage_service: set the topology kind value based on group 0 state`. I admit I got a bit lost when trying to organize the PR into commits and order them.

I'll try to reorganize the commits a bit and/or fix this piece of code so that it makes more sense.

gleb-cloudius replied:
Unless something substantial is broken by it at this point that prevents test from running no need to reorder (may be just fix the comment and in the later patches re-fix it back).

gleb-cloudius commented on service/topology_coordinator.cc:
```diff
@@ -2349,7 +2478,13 @@ future<> topology_coordinator::run() {
             }
 
+            if (_topo_sm._topology.ustate != topology::upgrade_state::done) {
+                co_await do_upgrade_step(std::move(guard));
```
So I wonder is it OK to start `cdc_generation_publisher_fiber()` and `start_tablet_load_stats_refresher()` above before upgrade is finished. May be the monitor fiber can run different functions based on `_topo_sm._topology.ustate` state `topology__coordinator::run` vs `topology__coordinator::run_upgrade`?

BTW I would rename `ustate` to `upgrade_state`

piodul replied:
> So I wonder is it OK to start `cdc_generation_publisher_fiber()` and `start_tablet_load_stats_refresher()` above before upgrade is finished.

Good point, it definitely sounds safer to start them only after upgrade is finished. There is a similar issue with `start_tablet_split_monitor` and `sstable_cleanup_fiber` that are started in `track_upgrade_progress_to_topology_coordinator`, I'll fix them, too.

> May be the monitor fiber can run different functions based on `_topo_sm._topology.ustate` state `topology__coordinator::run` vs `topology__coordinator::run_upgrade`?

Hmm yeah, I could introduce `topology_coordinator::run_upgrade` and call it in `run_topology_coordinator`. I'll try to reuse the error handling logic from `topology_coordinator::run()`.

> BTW I would rename `ustate` to `upgrade_state`

I followed the convention of the existing `tstate`, but makes sense to make `ustate` more descriptive. The only problem is that we already have an enum called `upgrade_state`. I suppose I could rename `upgrade_state` to `upgrade_state_type`?

gleb-cloudius replied:
Yes, rename it to _type. tstate is also not that great.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -6309,6 +6326,22 @@ future<> storage_service::wait_for_normal_state_handled_on_boot() {
+void storage_service::sync_topology_change_kind_with_group0() {
+    switch (_topology_state_machine._topology.ustate) {
+    case topology::upgrade_state::done:
+        _topology_change_kind_enabled = topology_change_kind::raft;
```
Why hide the assignment in the function? IMO better to have `topology_change_kind upgrate_state_to_topology_op_kind(upgrade_state)`.

piodul replied:
Sure, I can change it.

gleb-cloudius commented on service/storage_service.hh:
```diff
@@ -761,6 +761,17 @@ private:
+    // FIXME: This boolean flag is mostly needed because, shortly after starting
+    // a new group 0, the state of `system.topology` is empty, and updating the
+    // `_topology_change_kind_enabled` variable from group 0 state would lead
+    // to wrong results. This could be fixed by writing an initial `system.topology`
```
Do we even need `_topology_change_kind_enabled` in addition to `upgrade_state` in the topology? Why can't we have a function that returns `kind` depending on upgrade_state. It will also know how to handle empty topology.

piodul replied:
One issue is this FIXME. Another issue is that it affects gossip callbacks in storage_service, and gossiping starts before joining group 0 - we need these callbacks to operate in raft topology mode before we join a cluster that operates in raft topology mode.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -398,7 +398,11 @@ future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tm
+                { gms::application_state::TOKENS, gms::versioned_value::tokens(rs.ring.value().tokens) },
+                { gms::application_state::CDC_GENERATION_ID, gms::versioned_value::cdc_generation_id(_topology_state_machine._topology.current_cdc_generation_id) },
+                { gms::application_state::STATUS, gms::versioned_value::normal(rs.ring.value().tokens) }
+            });
```
It's a step in the wrong direction. We should have been removing things from gossiper not adding them back. Even having a state here is a hack :(

kbr-scylla replied:
Once we'll implement another kind of recovery (using external tool to calculate new topology instead of going through gossip-based removenode), we'll be able to get rid of this.

gleb-cloudius commented on main.cc:
```diff
@@ -1806,7 +1806,7 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             // storage proxy's and migration manager's verbs may access group0.
             // This will also disable migration manager schema pulls if needed.
-            group0_service.setup_group0_if_exist(sys_ks.local(), ss.local(), qp.local(), mm.local(), raft_topology_change_enabled).get();
+            group0_service.setup_group0_if_exist(sys_ks.local(), ss.local(), qp.local(), mm.local(), false).get();
```
If this param is now always false it should be removed.

gleb-cloudius reviewed (commented):
Left some comment. Sadly there is quite a bit of code to switch between nodes for recovery purposes. I honestly am not sure how relevant this recovery mode will be with tablets and we want to move to tablets together with moving to the raft topology.

kbr-scylla commented on docs/dev/topology-over-raft.md:
```diff
@@ -405,3 +411,72 @@ In case of failures like timeouts, connection issues, etc., the topology
+After going out of the `not_upgraded` state, the current raft leader starts its
+topology coordinator fiber and coordinates the remaining steps:
+
+- `build_coordinator_state` - the topology coordinator reads the contents
```
A global barrier at this point could also serve as a connectivity check @gleb-cloudius @piodul.

I believe we shouldn't progress upgrade beyond this step if we don't verify that Raft commands can reach every node.

This is not something that the administrator can verify easily.

kbr-scylla replied:
On the other hand, if the administrator is responsible for checking that every node reached `done`, if this wait "times out", they'll have to do disaster recovery anyway. So ok, as long as the admin is responsible for monitoring the state.

kbr-scylla commented:
> Left some comment. Sadly there is quite a bit of code to switch between nodes for recovery purposes. I honestly am not sure how relevant this recovery mode will be with tablets and we want to move to tablets together with moving to the raft topology.


It shouldn't become any less relevant than it is. After all, the purpose is to recover from majority loss, tablets or not majority loss can happen all the same.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -2457,6 +2454,50 @@ future<> storage_service::join_cluster(sharded<db::system_distributed_keyspace>&
+        _topology_change_kind_enabled = topology_change_kind::raft;
+    } else if (_sys_ks.local().bootstrap_complete()) {
+        // We already bootstrapped but we are not a part of group 0. This means that we are restarting after recovery.
+        slogger.info("Restarting after recovery - starting in legacy topology operations mode");
```
Could be restarting with upgrade (5.4 with Raft disabled -> 6.0) not just recovery

gleb-cloudius commented:
> > Left some comment. Sadly there is quite a bit of code to switch between nodes for recovery purposes. I honestly am not sure how relevant this recovery mode will be with tablets and we want to move to tablets together with moving to the raft topology.
> 
> It shouldn't become any less relevant than it is. After all, the purpose is to recover from majority loss, tablets or not majority loss can happen all the same.

Yes, but the complexity is  because you want to server traffic while recovering. Otherwise you couldn't care less about reading CDC from the correct table for instance. Tablets will not be available in recovery mode.

kbr-scylla commented on service/topology_coordinator.cc:
```diff
@@ -2236,6 +2240,131 @@ future<> topology_coordinator::start_tablet_load_stats_refresher() {
+        }
+
+        const auto& epmap = eptr->get_application_state_map();
+        auto get_application_state = [&] (gms::application_state app_state) -> sstring {
```
Could also be used in `get_sharding_info_for_host_id` above

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -1616,6 +1616,74 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
+        _sstable_cleanup_fiber = sstable_cleanup_fiber(_group0->group0_server(), proxy);
+        start_tablet_split_monitor();
+    } catch (...) {
+        // The calls above can theoretically fail due to coroutine frame allocation failure.
```
coroutine frame allocations may already be aborting in seastar, not sure

piodul replied:
https://en.cppreference.com/w/cpp/language/coroutines#Dynamic_allocation

We don't define `get_return_object_on_allocation_failure` in seastar. The `std::bad_alloc` exception is propagated to the caller. We do abort when we fail to allocate space for a `.then` continuation, though.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -1114,6 +1114,66 @@ future<> storage_service::update_topology_with_local_metadata(raft::server& raft
+        if (_topology_state_machine._topology.ustate != topology::upgrade_state::not_upgraded) {
+            co_return;
+        }
+        if (_topology_state_machine._topology.ustate == topology::upgrade_state::done) {
```
We'll never enter this branch

kbr-scylla commented on service/raft/group0_state_machine.cc:
```diff
@@ -43,11 +43,34 @@
+            // and will decide not to use `raft_pull_topology_snapshot` verb.
+            //
+            // The above issue is mitigated by requiring administrators to
+            // wait for the duration of ring delay before starting the upgrade
```
They should verify the feature is enabled (waiting as long as it takes), not wait for ring_delay.

kbr-scylla commented on cdc/generation.cc:
```diff
@@ -463,22 +463,44 @@ static std::optional<cdc::generation_id> get_generation_id_for(const gms::inet_a
     return gms::versioned_value::cdc_generation_id_from_string(gen_id_string);
 }
 
+static future<std::optional<cdc::topology_description>> retrieve_generation_data_v2(
```
A test which writes to CDC-enabled table while doing recovery might be useful.

Let's do it in a follow-up

kbr-scylla commented on test/topology_experimental_raft/test_topology_recovery_basic.py:
```diff
@@ -0,0 +1,77 @@
+        wait_until_topology_upgrade_finishes, delete_raft_topology_state, check_system_topology_and_cdc_generations_v3_consistency
+
+
+@pytest.mark.asyncio
```
Please run the new tests 100 times in debug mode

piodul replied:
They are currently running here: https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/2059/ . Will check the results tomorrow.

piodul replied:
Sigh, there was one crash.

A test was setting a 3-node cluster and, when the third node was being added, the second one crashed:

```
INFO  2024-02-08 21:47:54,586 [shard 0:strm] cdc - Started using generation (2024/02/08 19:47:54, f363ae60-c6ba-11ee-92a9-279f76963638).
DEBUG 2024-02-08 21:47:54,605 [shard 0: gms] raft_topology - topology cmd rpc barrier_and_drain is called
DEBUG 2024-02-08 21:47:54,606 [shard 0:strm] raft_topology - reload raft topology state
DEBUG 2024-02-08 21:47:54,624 [shard 0:strm] raft_topology - topology_state_load: current CDC generation ID: (2024/02/08 19:47:54, f363ae60-c6ba-11ee-92a9-279f76963638)
DEBUG 2024-02-08 21:47:54,625 [shard 0: gms] raft_topology - Got raft_topology_cmd::barrier_and_drain, version 6, current version 6
DEBUG 2024-02-08 21:47:54,625 [shard 0: gms] raft_topology - raft_topology_cmd::barrier_and_drain done
DEBUG 2024-02-08 21:47:54,626 [shard 1: gms] raft_topology - Got raft_topology_cmd::barrier_and_drain, version 6, current version 6
DEBUG 2024-02-08 21:47:54,626 [shard 1: gms] raft_topology - raft_topology_cmd::barrier_and_drain done
DEBUG 2024-02-08 21:47:54,658 [shard 0:strm] raft_topology - reload raft topology state
DEBUG 2024-02-08 21:47:54,682 [shard 0:strm] raft_topology - topology_state_load: current CDC generation ID: (2024/02/08 19:47:54, f363ae60-c6ba-11ee-92a9-279f76963638)
ERROR 2024-02-08 21:47:54,709 [shard 1:main] database - regular column id 0 >= 0 ...
```

I see there is an issue already which, superficially, looks similar: https://github.com/scylladb/scylladb/issues/16821 . There, the crash also happens when adding a node in raft topology mode, the error message is similar (most likely crashes in the same place) - however it happens during repair, and here... I don't know where it happened, there is a backtrace but it's not decoded.

Will investigate further tomorrow (for real this time).

piodul replied:
Yeah, I'm pretty sure that it's #16821.

- The test is known to be flaky
- The failure happens at the beginning of the test, when a cluster in raft mode is being set up. The PR slightly changes how new nodes join the cluster (they do raft discovery earlier and there is an additional RPC to ask the cluster about which mode to use), but judging by the logs it looks like it works correctly. The crash only happens during streaming during bootstrap, which should not be affected by this PR.

kbr-scylla commented:
> Tablets will not be available in recovery mode.

Only because we didn't implement tablets in recovery. We should consider this
@avikivity @tgrabiec

kbr-scylla commented:
> Only because we didn't implement tablets in recovery. We should consider this

I mean, without load balancing -- but last known replica sets could be loaded from `system.tablets`

avikivity commented:
What does that mean? Isn't the cluster generally unavailable during recovery?

avikivity commented:
Eventually, we want to be able to upgrade automatically (this is so that we can retire the gossip topology code). But it doesn't have to be now.

kbr-scylla commented:
> What does that mean? Isn't the cluster generally unavailable during recovery?

no, (vnode-based) queries are working if you have the replicas

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6534/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6534/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7775/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6534/)
#### Build Details:
- Duration: 2 hr 18 min
- Builder: spider1.cloudius-systems.com

piodul commented:
v3:

- Removed `get_sys_ks()` that returned a shared_ptr, now the code that was changed in `generation.cc` uses a reference to the system keyspace
- Renamed `ustate` to `upgrade_state`
- Removed the last argument from `setup_group0_if_exist` which was always set to false
- Removed `ustate == topology::upgrade_state::done` check from `start_upgrade_to_raft_topology` which was dead code
- `sync_topology_change_kind_with_group0` removed and replaced with assignments `_topology_change_kind_enabled = upgrade_state_to_topology_op_kind(...)`
- Topology coordinator-adjacent fibers are now started only after upgrade to raft topology has completed
- Various fixes in comments and log messages

kbr-scylla commented on service/topology_coordinator.cc:
```diff
@@ -2239,6 +2245,132 @@ future<> topology_coordinator::start_tablet_load_stats_refresher() {
+    rtlogger.info("waiting for all nodes to finish upgrade to raft schema");
+    release_guard(std::move(guard));
+    co_await _group0.wait_for_all_nodes_to_finish_upgrade(_as);
+    guard = co_await start_operation();
```
Ah, no, `start_operation()` is checking the term to prevent that. Nevermind
